compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs

```diff
@@ -197,38 +530,146 @@ impl<'tcx> WrongNumberOfGenericArgs<'_, 'tcx> {
+
+        // we first try to get lifetime name suggestions from scope or elision information. If none is
+        // available we use the parameter defintions
+        let suggested_args = if let Some(hir_id) = self.path_segment.hir_id {
+            if let Some(lifetimes_in_scope) = self.tcx.lifetime_scope(hir_id) {
+                match lifetimes_in_scope {
+                    LifetimeScopeForPath::NonElided(param_names) => {
+                        debug!("NonElided(param_names: {:?})", param_names);
+                        let p_names = Lrc::clone(&param_names);
+
+                        if p_names.len() >= num_params_to_take {
+                            // use lifetime parameters in scope for suggestions
+                            p_names
+                                .iter()
+                                .take(num_params_to_take)
+                                .map(|p| (*p).clone())
+                                .collect::<Vec<_>>()
+                                .join(", ")
+                        } else {
+                            // Not enough lifetime arguments in scope -> create suggestions from
+                            // lifetime parameter names in definition. An error for the incorrect
+                            // lifetime scope will be output later.
+                            self.get_lifetime_args_suggestions_from_param_names(num_params_to_take)
+                        }
+                    }
+                    LifetimeScopeForPath::Elided => {
+                        debug!("Elided");
+                        // use suggestions of the form `<'_, '_>` in case lifetime can be elided
+                        ["'_"].repeat(num_params_to_take).join(",")
+                    }
+                }
+            } else {
+                self.get_lifetime_args_suggestions_from_param_names(num_params_to_take)
+            }
         } else {
-            let span =
-                self.gen_args.args[self.args_offset + self.provided - 1].span().shrink_to_hi();
-            (span, ", ")
+            self.get_lifetime_args_suggestions_from_param_names(num_params_to_take)
         };
 
-        let msg = format!("add missing {} argument{}", self.kind, pluralize!(missing_arg_count));
+        debug!("suggested_args: {:?}", &suggested_args);
 
-        let sugg = self
-            .gen_params
-            .params
-            .iter()
-            .skip(self.params_offset + self.provided)
-            .take(missing_arg_count)
-            .map(|param| param.name.to_string())
-            .collect::<Vec<_>>()
-            .join(", ");
+        match self.angle_brackets {
+            AngleBrackets::Missing => {
+                let span = self.path_segment.ident.span;
+
+                // insert a suggestion of the form "Y<'a, 'b>"
+                let ident = self.path_segment.ident.name.to_ident_string();
+                let sugg = ident + "<" + &suggested_args + ">";
```

Comment:
Any reason not to use `format`?

```suggestion
                let sugg = format!("{}<{}>", self.path_segment.ident.name, suggested_args);
```

Reply:
No. I'll change that, thanks.